subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java

```diff
@@ -0,0 +1,140 @@
+package hu.bme.mit.theta.core.type.fptype;
+
+import hu.bme.mit.theta.common.Utils;
+import hu.bme.mit.theta.core.type.Expr;
+import hu.bme.mit.theta.core.type.abstracttype.AddExpr;
+import hu.bme.mit.theta.core.type.abstracttype.Additive;
+import hu.bme.mit.theta.core.type.abstracttype.DivExpr;
+import hu.bme.mit.theta.core.type.abstracttype.EqExpr;
+import hu.bme.mit.theta.core.type.abstracttype.Equational;
+import hu.bme.mit.theta.core.type.abstracttype.GeqExpr;
+import hu.bme.mit.theta.core.type.abstracttype.GtExpr;
+import hu.bme.mit.theta.core.type.abstracttype.LeqExpr;
+import hu.bme.mit.theta.core.type.abstracttype.LtExpr;
+import hu.bme.mit.theta.core.type.abstracttype.MulExpr;
+import hu.bme.mit.theta.core.type.abstracttype.Multiplicative;
+import hu.bme.mit.theta.core.type.abstracttype.NegExpr;
+import hu.bme.mit.theta.core.type.abstracttype.NeqExpr;
+import hu.bme.mit.theta.core.type.abstracttype.Ordered;
+import hu.bme.mit.theta.core.type.abstracttype.PosExpr;
+import hu.bme.mit.theta.core.type.abstracttype.SubExpr;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FpType implements Equational<FpType>, Additive<FpType>, Multiplicative<FpType>, Ordered<FpType> {
+	private final static int HASH_SEED = 5424;
+	private final static String TYPE_LABEL = "Fp";
+
+	private final int exponent;
+	private final int significand;
+
+	private volatile int hashCode = 0;
+
+	private FpType(final int exponent, final int significand) {
+		checkArgument(exponent > 1);
+		checkArgument(significand > 1);
+		this.exponent = exponent;
+		this.significand = significand;
+	}
+
+	public static FpType of(final int exponent, final int significand) {
+		return new FpType(exponent, significand);
+	}
+
+	public int getExponent() {
+		return exponent;
+	}
+
+	public int getSignificand() {
+		return significand;
+	}
+
+	@Override
+	public EqExpr<FpType> Eq(Expr<FpType> leftOp, Expr<FpType> rightOp) {
+		return FpEqExpr.of(leftOp, rightOp);
+	}
+
+	@Override
+	public NeqExpr<FpType> Neq(Expr<FpType> leftOp, Expr<FpType> rightOp) {
+		return FpNeqExpr.of(leftOp, rightOp);
+	}
+
+	@Override
+	public int hashCode() {
+		int result = hashCode;
+		if (result == 0) {
+			result = HASH_SEED;
+			result = 31 * result + exponent;
+			result = 31 * result + significand;
+			hashCode = result;
+		}
+		return result;
+	}
+
+	@Override
+	public boolean equals(final Object obj) {
+		if (this == obj) {
+			return true;
+		} else if (obj instanceof FpType) {
+			final FpType that = (FpType) obj;
+			return this.getExponent() == that.getExponent() && this.getSignificand() == that.getSignificand();
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public String toString() {
+		return Utils.lispStringBuilder(TYPE_LABEL).add(exponent).add(significand).toString();
+	}
+
+	@Override
+	public AddExpr<FpType> Add(Iterable<? extends Expr<FpType>> ops) {
+		return FpExprs.Add(FpRoundingMode.getDefaultRoundingMode(), ops);
```

Comment:
This default rounding mode has some implications and I am not sure whether this is intentional. This function is called by the grammar mainly to create the expressions it sees. So this line of code (and the other similarly) mean that you can only create expressions of one specific rounding type through the grammar.

Reply:
Somewhat fixed in [2608512](https://github.com/ftsrg/theta/pull/123/commits/260851260a768790fd16a6f98dcd612f29bb0084) -- see separate discussion